File: test/reference.test.ts
import { describe, expect, test } from "bun:test";
import {
  CAPABILITIES,
  capabilities,
  missingCapabilities,
  requireCapabilities,
  type UnsupportedCapabilityError,
} from "../src/domain.ts";
import type { RefKind } from "../src/reference.ts";
import { parseRef, workspaceRefOf } from "../src/reference.ts";

describe("parseRef", () => {
  test.each([
    ["jira:responsibid", "tracker"],
    ["jira:responsibid:JWB", "workspace"],
    ["jira:responsibid:JWB:group:JWB-150", "group"],
    ["jira:responsibid:JWB:map:JWB-239", "map"],
    ["jira:responsibid:JWB:ticket:JWB-245", "ticket"],
    ["wayfinder-run:018f", "run"],
    ["wayfinder-claim:018f", "claim"],
  ] as const)("parses %s", (raw, kind) => expect(parseRef(raw).kind).toBe(kind as RefKind));

  test.each([
    "JWB-245",
    "jira:",
    "jira:x:y:map",
    "jira:x:y:other:z",
    "wayfinder-run:",
    "nav-run:018f",
    "nav-claim:018f",
  ])("rejects %s", (raw) => expect(() => parseRef(raw)).toThrow());
});

test("derives the same qualified workspace from workspace-owned references", () => {
  expect(workspaceRefOf("jira:responsibid:JWB")).toBe("jira:responsibid:JWB");
  expect(workspaceRefOf("jira:responsibid:JWB:group:JWB-100")).toBe("jira:responsibid:JWB");
  expect(workspaceRefOf("jira:responsibid:JWB:map:JWB-239")).toBe("jira:responsibid:JWB");
  expect(workspaceRefOf("jira:responsibid:JWB:ticket:JWB-245")).toBe("jira:responsibid:JWB");
  expect(() => workspaceRefOf("jira:responsibid")).toThrow("not owned by a workspace");
});

test("reports missing capabilities", () => {
  expect(
    missingCapabilities(capabilities("native_maps"), capabilities("native_maps", "claim_comments")),
  ).toEqual(["claim_comments"]);
});

test("capability vocabulary is unique and protocol-safe", () => {
  expect(new Set(CAPABILITIES).size).toBe(CAPABILITIES.length);
  for (const capability of CAPABILITIES) {
    expect(capability).toMatch(/^[a-z][a-z0-9]*(?:_[a-z0-9]+)*$/);
  }
});

test("unsupported capabilities fail explicitly with a stable code", () => {
  expect(() =>
    requireCapabilities(
      capabilities("native_maps"),
      capabilities("native_maps", "conditional_update", "claim_comments"),
    ),
  ).toThrow(
    expect.objectContaining({
      name: "UnsupportedCapabilityError",
      code: "unsupported_capability",
      missing: ["conditional_update", "claim_comments"],
    }) as UnsupportedCapabilityError,
  );
});

File: internal/domain/reference_test.go
package domain

import "testing"

func TestParseRef(t *testing.T) {
	tests := []struct {
		raw  string
		kind RefKind
	}{
		{"jira:responsibid", RefTracker},
		{"jira:responsibid:JWB", RefWorkspace},
		{"jira:responsibid:JWB:group:JWB-150", RefGroup},
		{"jira:responsibid:JWB:map:JWB-239", RefMap},
		{"jira:responsibid:JWB:ticket:JWB-245", RefTicket},
		{"nav-run:018f", RefRun}, {"nav-claim:018f", RefClaim},
	}
	for _, test := range tests {
		parsed, err := ParseRef(test.raw)
		if err != nil {
			t.Fatalf("%s: %v", test.raw, err)
		}
		if parsed.Kind != test.kind {
			t.Errorf("%s: got %s", test.raw, parsed.Kind)
		}
	}
}

func TestParseRefRejectsAmbiguousValues(t *testing.T) {
	for _, raw := range []string{"JWB-245", "jira:", "jira:x:y:map", "jira:x:y:other:z", "nav-run:"} {
		if _, err := ParseRef(raw); err == nil {
			t.Errorf("expected %q to fail", raw)
		}
	}
}

func TestCapabilitiesMissing(t *testing.T) {
	have := NewCapabilities("a", "b")
	missing := have.Missing(NewCapabilities("b", "c"))
	if len(missing) != 1 || missing[0] != "c" {
		t.Fatalf("unexpected missing set: %v", missing)
	}
}
